File: backend/routers/nodes.py
"""
知识节点路由
"""

from fastapi import APIRouter, Depends, HTTPException
from models import NodeUpdate, NodeReviewResult, NodeSchema, NodeBrief
from services import MemoryService, NodeService
from dependencies import get_current_user

router = APIRouter(prefix="/api/nodes", tags=["nodes"])


@router.get("", response_model=list[NodeBrief])
def list_nodes(user_id: str = Depends(get_current_user)):
    return NodeService().list_nodes(user_id)


@router.get("/due", response_model=list[NodeBrief])
def due_today(user_id: str = Depends(get_current_user)):
    return NodeService().list_due_today(user_id)


@router.get("/{node_id}", response_model=NodeSchema)
def get_node(node_id: str, user_id: str = Depends(get_current_user)):
    node = NodeService().get_node(user_id, node_id)
    if not node:
        raise HTTPException(404, "节点不存在")
    return node


@router.patch("/{node_id}", response_model=NodeSchema)
def update_node(
    node_id: str,
    data: NodeUpdate,
    user_id: str = Depends(get_current_user),
):
    node = NodeService().update_node(user_id, node_id, data)
    if not node:
        raise HTTPException(404, "节点不存在")
    return node


@router.delete("/{node_id}", status_code=204)
def delete_node(node_id: str, user_id: str = Depends(get_current_user)):
    ok = NodeService().delete_node(user_id, node_id)
    if not ok:
        raise HTTPException(404, "节点不存在")


@router.post("/{node_id}/review", response_model=dict)
def review_node(
    node_id: str,
    body: NodeReviewResult,
    user_id: str = Depends(get_current_user),
):
    try:
        return MemoryService().review_node_for_user(user_id, node_id, body.quality)
    except ValueError as e:
        raise HTTPException(404, str(e))

File: frontend/src/api/nodes.js
/**
 * 知识节点 API
 * GET    /api/nodes          所有节点（图谱数据，含 relations）
 * GET    /api/nodes/due      今日到期复习节点
 * GET    /api/nodes/:id      节点详情（含 content / origins）
 * POST   /api/nodes          创建节点
 * PATCH  /api/nodes/:id      编辑节点
 * DELETE /api/nodes/:id      删除节点
 * POST   /api/nodes/:id/review  提交复习结果（SM-2）
 */
import { api } from "./base.js";

export const nodeApi = {
  // 获取所有节点（用于知识图谱、知识库）
  getAll: () => api.get("/api/nodes"),

  // 获取今日到期复习节点
  getDue: () => api.get("/api/nodes/due"),

  // 获取单个节点详情
  getById: (id) => api.get(`/api/nodes/${id}`),

  // 创建节点
  create: (data) => api.post("/api/nodes", data),

  // 手动更新节点信息（标题、描述、标签等）
  update: (id, data) => api.patch(`/api/nodes/${id}`, data),

  // 删除节点
  delete: (id) => api.delete(`/api/nodes/${id}`),

  // 提交复习结果，触发 SM-2 更新
  // quality: 0~5（5=完全记住，0=完全忘了）
  review: (id, quality) => api.post(`/api/nodes/${id}/review`, { quality }),
};
